### PR TITLE
Handle errors during error recovery gracefully

### LINKED ISSUE #62546: internal compiler error: Error constructed but not emitted
I see the following error on the `beta` and `nightly` channels, but not on `stable`:
```
$ echo "pub t(#" > main.rs

$ rustc main.rs
error: this file contains an un-closed delimiter
 --> main.rs:1:9
  |
1 | pub t(#
  |      -  ^
  |      |
  |      un-closed delimiter

error: internal compiler error: Error constructed but not emitted

thread 'rustc' panicked at 'explicit panic', src/librustc_errors/diagnostic_builder.rs:387:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to 2 previous errors


error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-beta.2 (74e5a0d47 2019-07-08) running on x86_64-apple-darwin
```

Found with the help of [libfuzzer-sys](https://github.com/rust-fuzz/libfuzzer-sys).

### PR DESCRIPTION
**Pull request by estebank**
Fix #62546.

### CODE REVIEW

**Comment by estebank** [comment]
cc @petrochenkov

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -7408,10 +7408,13 @@ impl<'a> Parser<'a> {
...
                 self.bump();  // `(`
-                let kw_name = if let Ok(Some(_)) = self.parse_self_arg_with_attrs() {
-                    "method"
-                } else {
-                    "function"
+                let kw_name = match self.parse_self_arg_with_attrs() {
+                    Ok(Some(_)) => "method",
+                    Ok(None) => "function",
+                    Err(mut err) => {
+                        err.cancel();
```
Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: You could also use `.map_err(|e| e.cancel())`.

**Review by eddyb — APPROVED** [review]
r=me

**Comment by Mark-Simulacrum** [comment]
I noted that @estebank mentioned in #62637 that this might warrant a stable point release. The release team has no current plans to conduct one, and it is somewhat harder to do this cycle, so we'd like to avoid it if possible -- is this sufficiently high impact to warrant that? It _feels_ like it's not too severe to me? If you decide you would like to consider a stable release, please label this PR with the stable-nominated tag and ask compiler team to approve stable backport; if that all follows release team will discuss whether we would agree that this warrants a stable backport.

To summarize, stable point release process is as follows:
 * compiler team approves _code_ as being safe and acceptable for backport to stable (this is same as with beta, but the bar for safety is higher, as we get into ecosystem hands much more quickly and with less testing)
 * release team discusses whether stable point release is warranted, and if so informs compilers team and proceeds

**Comment by estebank** [comment]
@Mark-Simulacrum I should have linked to PR #62607, not this PR. That PR is already nominated and should be discussed next thursday. I feel that PR is certainly bad but not enough to justify a point release.
